docs: Add "How to upgrade etcd" (#802)
kinvolk/lokomotive: docs/how-to-guides/upgrade-etcd.md

              show -p InvocationID --value etcd-member.service)
```

> **NOTE**: Do not proceed with the upgrade of the rest of the cluster if you encounter any errors.

Once you see the following log line, you can discern that the etcd daemon has come up without errors:

```log
etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]
```

Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:

```log
embed: serving client requests on 10.88.81.1:2379
```

### Step 5: Verify using `etcdctl`

We can use `etcdctl` client to verify the state of etcd cluster.

```bash
# Find the endpoint of this node's etcd:
export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | tr -d '"')

export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
              --cert=/etc/ssl/etcd/etcd-client.crt \
              --key=/etc/ssl/etcd/etcd-client.key \
              --endpoints=${endpoint}"

# Verify:
sudo ETCDCTL_API=3 etcdctl member list $flags
sudo ETCDCTL_API=3 etcdctl endpoint health $flags
```

> The last command should report that node is healthy. If it indicates otherwise then try commands from [Step 4](#step-4-verify-upgrade) to see what's wrong. If the node is healthy, it is safe to move forward with the next controller node.

Reviewer: Just curious: shouldn't the node be healthy and the first command indicate that there are 3 members? I mean, in normal scenarios if there are 2 members and the node is healthy (not sure this can happen), we wouldn't want to continue, right?

Author: > if there are 2 members and the node is healthy (not sure this can happen),

Do you mean unhealthy here?

Reviewer: No, healthy. I mean, let's say the last command `etcdctl endpoint health` shows healthy, but the previous one about members shows 2. We don't want to continue, right? Not sure if such a thing can happen, though, as if you are healthy you might have joined the cluster and already synced etcd data?

Author: @rata now we have a reliable way of finding all the endpoints \o/. See the current changes.

```bash
core@mycluster-controller-2 ~ $ # Find the endpoint of this node's etcd:
core@mycluster-controller-2 ~ $ export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS \
>         /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | tr -d '"')

core@mycluster-controller-2 ~ $ export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
>               --cert=/etc/ssl/etcd/etcd-client.crt \
>               --key=/etc/ssl/etcd/etcd-client.key"

core@mycluster-controller-2 ~ $ endpoints=$(sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint} \
>             --write-out=json | jq -r '.members[].clientURLs[]')

core@mycluster-controller-2 ~ $ endpoints=$(sed 's| |,|g' <<< ${endpoints})

core@mycluster-controller-2 ~ $ # Verify:
core@mycluster-controller-2 ~ $ sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint}
2b89a567f6c2e2e8, started, etcd0, https://mycluster-etcd0.dev.mycomp.com:2380, https://mycluster-etcd0.dev.mycomp.com:2379
99588ac1031498cf, started, etcd2, https://mycluster-etcd2.dev.mycomp.com:2380, https://mycluster-etcd2.dev.mycomp.com:2379
eec240697239d5b8, started, etcd1, https://mycluster-etcd1.dev.mycomp.com:2380, https://mycluster-etcd1.dev.mycomp.com:2379

core@mycluster-controller-2 ~ $ sudo ETCDCTL_API=3 etcdctl endpoint health $flags --endpoints=${endpoints}
https://mycluster-etcd2.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 29.811666ms
https://mycluster-etcd0.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 40.142713ms
https://mycluster-etcd1.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 40.819513ms
```

Reviewer: Cool!
My point still is:

> My point is: before continue we want to make sure the cluster has 3 members again and etcd data has been fully synced. Not sure if healthy implies all of that or not :) (maybe it is healthy because joined and is syncing, but haven't completed? Don't really know :))

So, I'd update the text to reflect that (if this is enough to see that etcd has 3 members and all data is synced) :)

Author: Data is not lost when you upgrade etcd, the data is already there. 

We can add something like this at the end and say that verify that the db size is same for all the nodes.

```console
$ sudo ETCDCTL_API=3 etcdctl endpoint status -w table $flags --endpoints=${endpoints}
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
|               ENDPOINT              |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
| https://mycluster-etcd0.mycomp:2379 | 2b89a567f6c2e2e8 |   3.4.7 |  7.6 MB |      true |         2 |      46614 |
| https://mycluster-etcd2.mycomp:2379 | 99588ac1031498cf |   3.4.7 |  7.6 MB |     false |         2 |      46614 |
| https://mycluster-etcd1.mycomp:2379 | eec240697239d5b8 |   3.4.7 |  7.7 MB |     false |         2 |      46614 |
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
```

Reviewer: Nah, seems too complicated. Also, same size doesn't mean it is in sync (a small downtime, like restaritng etcd on upgrade, won't be noticeable on mega bytes but can still missed a small change).

I guess this is fine for now. Thanks again! :)

```bash
sudo journalctl _SYSTEMD_INVOCATION_ID=$(sudo systemctl \
              show -p InvocationID --value etcd-member.service)
```

> **NOTE**: Do not proceed with the upgrade of the rest of the cluster if you encounter any errors.

Once you see the following log line, you can discern that the etcd daemon has come up without errors:

```log
etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]
```

Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:

```log
embed: serving client requests on 10.88.81.1:2379
```

### Step 5: Verify using `etcdctl`

We can use `etcdctl` client to verify the state of etcd cluster.

```bash
# Find the endpoint of this node's etcd:
export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS \
        /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | tr -d '"')
export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
              --cert=/etc/ssl/etcd/etcd-client.crt \
              --key=/etc/ssl/etcd/etcd-client.key"
endpoints=$(sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint} \

>             --write-out=json | jq -r '.members | map(.clientURLs) | add | join(",")')

Reviewer: IMHO, these `jq` lines are too complicared. But this also can be simplified later, if it is worth the effort :)

Reviewer: Marking as resolved as this was suggested by @invidian . Don't want to go into a loop here :-D

# Upgrading etcd

## Contents

- [Introduction](#introduction)
- [Steps](#steps)

>   - [Step 1: Find out the IP and SSH](#step-1-find-out-the-ip-and-ssh)

Reviewer 2: Find out the IP is not something we are doing, below makes a better title IMO.
```suggestion
  - [Step 1: SSH to the controller node ](#step-1-ssh-to-the-controller-node)
```

  - [Step 2: Create necessary directories with correct permissions](#step-2-create-necessary-directories-with-correct-permissions)
  - [Step 3: Upgrade etcd](#step-3-upgrade-etcd)
  - [Step 4: Verify upgrade](#step-4-verify-upgrade)
  - [Step 5: Verify using `etcdctl`](#step-5-verify-using-etcdctl)

## Introduction

[Etcd](https://etcd.io/) is the most crucial component of a Kubernetes cluster. It stores the cluster state.

This document will provide step by step guide on upgrading etcd in Lokomotive.

## Steps

Repeat the following steps on all the controller node one node at a time.

### Step 1: Find out the IP and SSH

Find the IP of the controller node by visiting the cloud provider dashboard and ssh into it.

```bash
ssh core@<IP Address>
```

### Step 2: Create necessary directories with correct permissions

> Latest etcd (`>= v3.4.10`) necessitates the data directory permissions to be `0700`, accordingly change the permissions. Verify the permissions are changed to `rwx------`.

Reviewer 2: It would be better if we provide a link to the upstream PR or release notes document.

```bash
sudo chmod 0700 /var/lib/etcd/
sudo ls -ld /var/lib/etcd/
```

If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes made above persistent run the following command:

```bash
echo "d    /var/lib/etcd 0700 etcd etcd - -" | sudo tee /etc/tmpfiles.d/etcd-wrapper.conf

> ```

Reviewer: This is already applied if we merge the other PR, right? I'd mention to do this *only* if using lokomotive <= 0.4.0 (or whatever)

> If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes made above persistent run the following command:

Reviewer 2: made above seems redundant.

```suggestion
If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes persistent run the following command:
```

```bash
export etcd_version=<latest etcd version e.g. v3.4.10>

sudo sed -i "s,ETCD_IMAGE_TAG=.*,ETCD_IMAGE_TAG=${etcd_version}," \
        /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
sudo systemctl daemon-reload
sudo systemctl restart etcd-member
```

### Step 4: Verify upgrade

Verify that the etcd service is in `active (running)` state:

```bash
sudo systemctl status --no-pager etcd-member
```

Run the following command to see logs of the process since the last restart:

```bash
sudo journalctl _SYSTEMD_INVOCATION_ID=$(sudo systemctl \
              show -p InvocationID --value etcd-member.service)
```

> **NOTE**: Do not proceed with the upgrade of the rest of the cluster if you encounter any errors.

Once you see the following log line, you can discern that the etcd daemon has come up without errors:

```log
etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]
```

Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:

```log
embed: serving client requests on 10.88.81.1:2379
```

> ### Step 5: Verify using `etcdctl`

Reviewer: Steps 1 to 4 seem to be needed for each node, however this step seems to be needed after finishing doing 1-4 for all nodes. Or is it something to do with every node?

Author: This is something that can be done on every node, now we don't use certs from bootkube dir but from the node's etcd certs dir.

Reviewer: Oh, cool! Thanks!

We can use `etcdctl` client to verify the state of etcd cluster.

> **NOTE**: Before proceeding to other commands, set the `no_of_controller_nodes` variable to the number of controller nodes in the cluster.

```bash
export no_of_controller_nodes=<no of controller nodes>

> # Find the endpoint of this node's etcd:

Reviewer: Of this node? Isn't this a list of the 3 endpoints? How is the code getting the one for this node?

Isn't currently serveral endpoint*s*?

Author: Earlier due to the misunderstanding I thought certs were only on controller-0, now that we know there are certs on each node we can only look at that particular node.

Reviewer: But is it doing something specific for this node? The code seems quite generic. Like, let's say you have 3 nodes, one down. Would the command say something about it?

Author: No it does say anything about other nodes.

Reviewer: Do have an example of the output of that command in a cluster with 3 controllers? I think it should include endpoints for the 3 of them.

Or am I missing something?

Author: > Or am I missing something?

I think I did a poor job of explaining what the problem was, let me try again.

> Of this node? Isn't this a list of the 3 endpoints? How is the code getting the one for this node?

```console
$ cat /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
[Unit]
Requires=create-etcd-config.service
After=create-etcd-config.service
[Service]
EnvironmentFile=/etc/kubernetes/etcd.config
Environment="ETCD_IMAGE_URL=docker://quay.io/coreos/etcd"
Environment="RKT_RUN_ARGS=--insecure-options=image"
Environment="ETCD_IMAGE_TAG=v3.4.7"
Environment="ETCD_NAME=etcd0"
Environment="ETCD_ADVERTISE_CLIENT_URLS=https://foobar-cluster-etcd0.k8s.com:2379"
Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=https://foobar-cluster-etcd0.k8s.com:2380"
Environment="ETCD_INITIAL_CLUSTER=etcd0=https://foobar-cluster-etcd0.k8s.com:2380,etcd1=https://foobar-cluster-etcd1.k8s.com:2380,etcd2=https://foobar-cluster-etcd2.k8s.com:2380"
Environment="ETCD_STRICT_RECONFIG_CHECK=true"
Environment="ETCD_SSL_DIR=/etc/ssl/etcd"
Environment="ETCD_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/server-ca.crt"
Environment="ETCD_CERT_FILE=/etc/ssl/certs/etcd/server.crt"
Environment="ETCD_KEY_FILE=/etc/ssl/certs/etcd/server.key"
Environment="ETCD_CLIENT_CERT_AUTH=true"
Environment="ETCD_PEER_TRUSTED_CA_FILE=/etc/ssl/certs/etcd/peer-ca.crt"
Environment="ETCD_PEER_CERT_FILE=/etc/ssl/certs/etcd/peer.crt"
Environment="ETCD_PEER_KEY_FILE=/etc/ssl/certs/etcd/peer.key"
Environment="ETCD_PEER_CLIENT_CERT_AUTH=true"
Environment=""
ExecStopPost=-/opt/etcd-rejoin
```

Each etcd conf file has a list of all servers but on the peer port as in `2380`. While when we as client talk to this etcd server on port `2379`.

Earlier due to my misunderstanding of the certificates not being on all nodes I suggested the user(in the doc) to run this particular last step only on the first controller node. But later I found that the certs are available on all the nodes.

Now the earlier script I had was written such that it ran correctly only on the controller-0. It use to extract the endpoint of controller-0 and then generate a list by replacing the string `etcd0`. Now the same script won't work on all nodes because on controller-1 the endpoint has etcd1 and so on. Now the script there for user to copy and paste was becoming very complicated so I removed all that and now ask user to only check the health of the node that user is on.

Another way to do is to get the list of peers from the service file do replace of 2380 by 2379 and then remove things like etcd0=, etc1=, etc. I tried this approach as well, but again the complexity of it was defeating the purpose of keeping the document simplistic.

> Isn't currently serveral endpoints?

There are several endpoints in the service file but they are listed as shown in the file above.

> But is it doing something specific for this node? The code seems quite generic. Like, let's say you have 3 nodes, one down. Would the command say something about it?

The command tells you about all the endpoints we give it. We just have to include comma separated list of all the endpoints. But as explained above I have changed it to include only the current node.

> Do have an example of the output of that command in a cluster with 3 controllers? I think it should include endpoints for the 3 of them.

```console
$ sudo ETCDCTL_API=3 etcdctl member list $flags
2b89a567f6c2e2e8, started, etcd0, https://foobar-cluster-etcd0.k8s.com:2380, https://foobar-cluster-etcd0.k8s.com:2379
99588ac1031498cf, started, etcd2, https://foobar-cluster-etcd2.k8s.com:2380, https://foobar-cluster-etcd2.k8s.com:2379
eec240697239d5b8, started, etcd1, https://foobar-cluster-etcd1.k8s.com:2380, https://foobar-cluster-etcd1.k8s.com:2379
```

```console
$ sudo ETCDCTL_API=3 etcdctl endpoint health $flags
https://foobar-cluster-etcd0.k8s.com:2379 is healthy: successfully committed proposal: took = 30.131317ms
```

Reviewer: Ohhh, I see. Sorry, I think that was partly my fault because the actual code line was not selected on my comment and that changed :facepalm: .

I now understand. I don't have all the details in mind about etcdctl, but I trust you and seems reasonable to me.

Thanks again for your patience! :)

export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
              --cert=/etc/ssl/etcd/etcd-client.crt \
              --key=/etc/ssl/etcd/etcd-client.key"
endpoints=$(sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint} \

>             --write-out=json | jq -r '.members[].clientURLs[]')

Reviewer: Isn't this run on the server? Is `jq` available on Flatcar? :thinking:

Author: jq is available on flatcar.

> endpoints=$(sed 's| |,|g' <<< ${endpoints})

Reviewer 3: `jq` also supports formatting, I think this set can be skipped.

Author: That will add a trailing comma. like `server1,server2,server3,` and etcdctl does not play well with extra comma. So we will still need another command to replace the last comma.

Reviewer 3: But it seems it also supports `| join(",")` :smile:

> endpoints=$(sed 's| |,|g' <<< ${endpoints})

Reviewer: I think ```<<<``` is a bashism, as long as we have bash on Flatcar by default for the core user will work, though. So not a big issue. Wanted to double check that this has been tested on Flatcar and not just locally on your PC

Reviewer: This will just not be a bashism and is more readable, IMHO. (not tested :D)

```suggestion
endpoints=$(printf ${endpoints} | sed 's| |,|g')
```

Reviewer 3: @rata `jq` supports `join(",")`, there is no need for this sed at all :smile:

Reviewer: @invidian yeah, I see it. I just mentioned this in case `jq` is not on Flatcar but just on suraj's PC. Thanks! :)

Author: These commands are run on server and jq is available on flatcar.

# Upgrading etcd

## Contents

- [Introduction](#introduction)
- [Steps](#steps)
  - [Step 1: Find out the IP and SSH](#step-1-find-out-the-ip-and-ssh)
  - [Step 2: Create necessary directories with correct permissions](#step-2-create-necessary-directories-with-correct-permissions)
  - [Step 3: Upgrade etcd](#step-3-upgrade-etcd)
  - [Step 4: Verify upgrade](#step-4-verify-upgrade)
  - [Step 5: Verify using `etcdctl`](#step-5-verify-using-etcdctl)

## Introduction

[Etcd](https://etcd.io/) is the most crucial component of a Kubernetes cluster. It stores the cluster state.

This document will provide step by step guide on upgrading etcd in Lokomotive.

## Steps

Repeat the following steps on all the controller node one node at a time.

### Step 1: Find out the IP and SSH

Find the IP of the controller node by visiting the cloud provider dashboard and ssh into it.

```bash
ssh core@<IP Address>
```

### Step 2: Create necessary directories with correct permissions

> Latest etcd (`v3.4.10`) necessitates the data directory permissions to be `0700`, accordingly change the permissions. Verify the permissions are changed to `rwx------`.

Reviewer: Now it is 3.4.10, but this will be innacurate soon. Maybe replace it with "etcd >=3.4.10 ..." (was introduced in this release, right?)

  - [Step 5: Verify using `etcdctl`](#step-5-verify-using-etcdctl)

## Introduction

[Etcd](https://etcd.io/) is the most crucial component of a Kubernetes cluster. It stores the cluster state.

This document will provide step by step guide on upgrading etcd in Lokomotive.

## Steps

Repeat the following steps on all the controller node one node at a time.

### Step 1: Find out the IP and SSH

Find the IP of the controller node by visiting the cloud provider dashboard and ssh into it.

```bash
ssh core@<IP Address>
```

### Step 2: Create necessary directories with correct permissions

Latest etcd (`v3.4.10`) necessitates the data directory permissions to be `0700`, accordingly change the permissions. Verify the permissions are changed to `rwx------`.

```bash
sudo chmod 0700 /var/lib/etcd/
sudo ls -ld /var/lib/etcd/
```

If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes made above persistent run the following command:

```bash
echo "d    /var/lib/etcd 0700 etcd etcd - -" | sudo tee /etc/tmpfiles.d/etcd-wrapper.conf
```

> ### Step 3: Upgrade Kubelet

Reviewer 3: Kubelet ? :)

Run the following commands:

> **NOTE**: Before proceeding to other commands, set the `etcd_version` variable to the latest etcd version.

```bash
export etcd_version=<latest etcd version e.g. v3.4.10>

> sudo sed -i "s|$(grep ETCD_IMAGE_TAG /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf)|Environment=\"ETCD_IMAGE_TAG=${etcd_version}\"|g" /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf

Reviewer: wow, this seems complicated.

Can't we just do:

```
sed -i "s,ETCD_IMAGE_TAG=.*,ETCD_IMAGE_TAG=${etcd_version}"," /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
```

Author: This worked thanks Rodrigo :-)

> systemctl daemon-reload

Reviewer 3: Does those command works without `sudo`?

systemctl restart etcd-member
```

### Step 4: Verify upgrade

Verify that the etcd service is in `active (running)` state:

```bash
systemctl status --no-pager etcd-member
```

Run the following command to see logs of the process since the last restart:

```bash
journalctl _SYSTEMD_INVOCATION_ID=$(systemctl show -p InvocationID --value etcd-member.service)
```

Once you see the following log line, you can discern that the etcd daemon has come up without errors:

```log
etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]

> ```

Reviewer: Shouldn't we mention that the user should NOT proceed with another node until the errors are solved? If they do, the cluster will lose quorum and we are in a not so nice situation

Author: Added a note after journalctl.

Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:

```log
embed: serving client requests on 10.88.81.1:2379
```

### Step 5: Verify using `etcdctl`

We can use `etcdctl` client to verify the state of etcd cluster.

> > **NOTE**: These commands can only be run on the `*controller-0` of your cluster because the required client certificates are readily available on that node.

Reviewer: Really? They are not present on other nodes? This seems weird, right?

Author: No, we scp (using terraform) the configs only on the 1st controller and after that bootkube takes care of disbursing it from there.

> **NOTE**: Before proceeding to other commands, set the `no_of_controller_nodes` variable to the number of controller nodes in the cluster.

```bash
export no_of_controller_nodes=<no of controller nodes>

# Get the required certificates:

> cd /opt/bootkube/assets/tls/

Reviewer: Oh, is it because you are using the files from bootkube? I see now :)

Author: They are copied in a different directory but there are no client certs there:

```console
$ sudo ls /etc/ssl/etcd/etcd/
peer-ca.crt  peer.crt  peer.key  server-ca.crt  server.crt  server.key
```

Author: Oh hold on they are available here:

```console
$ sudo ls /etc/ssl/etcd/
etcd  etcd-client-ca.crt  etcd-client.crt  etcd-client.key
```

Reviewer: Makes sense, they should be in some part for all servers, I think :)

Author: Thanks, removed the note and now it can be run on each node.

export ETCDCTL_API=3
export cacert=etcd-client-ca.crt
export cert=etcd-client.crt
export key=etcd-client.key

# Find the endpoint of etcd0:

> export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | sed 's|"||g')

Reviewer: I think `tr -d '|'` is more readable than the sed thing
